dao-voting/src/main.sw

```diff
@@ -0,0 +1,295 @@
+    /// is_yes_vote - determines if you vote yes or no on the proposal
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The proposal id is out of range
+    /// - The vote amount is 0
+    /// - The vote amount is greater than the users deposited balance
+    /// - The given proposal is expired
+    #[storage(read, write)]fn vote(proposal_id: u64, vote_amount: u64, is_yes_vote: bool) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(proposal_id < storage.proposal_count, Error::InvalidId);
+        require(vote_amount > 0, Error::VoteAmountCannotBeZero);
+
+        let sender: Identity = sender_identity();
+        let sender_balance = storage.balances.get(sender);
+
+        require(sender_balance >= vote_amount, Error::NotEnoughAssets);
+
+        let mut proposal = storage.proposals.get(proposal_id);
+        require(proposal.end_height >= height(), Error::ProposalExpired);
+
+        if (is_yes_vote) {
+            proposal.yes_votes = proposal.yes_votes + vote_amount;
+        } else {
+            proposal.no_votes = proposal.no_votes + vote_amount;
+        };
+
+        storage.balances.insert(sender, sender_balance - vote_amount);
+
+        let votes = storage.votes.get((sender, proposal_id));
+        storage.votes.insert((sender, proposal_id), votes + vote_amount);
+
+        storage.proposals.insert(proposal_id, proposal);
+
+        true
+    }
+
+    /// Execute a given proposal
+    ///
+    /// # Parameters
+    ///
+    /// proposal_id - proposal to execute
+    ///
+    /// # Panics
+    ///
+    /// This function will panic when:
+    /// - The construct has not been called to initialize
+    /// - The proposal id is out of range
+    /// - The proposal has not expired
+    /// - The proposal has not met the necessary approval percentage
+    #[storage(read, write)]fn execute(proposal_id: u64) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(proposal_id < storage.proposal_count, Error::InvalidId);
+
+        let proposal = storage.proposals.get(proposal_id);
+        require(proposal.end_height < height(), Error::ProposalActive);
+
+        // TODO figure out how to prevent approval percentage from overflowing
+        // When close to the u64 max
+        // https://github.com/FuelLabs/sway-applications/issues/106
+        let approval_percentage = proposal.yes_votes * 100 / (proposal.yes_votes + proposal.no_votes);
+        require(approval_percentage >= proposal.approval_percentage, Error::ApprovalPercentageNotMet);
+
+        asm(rA: proposal.call_data.memory_address, rB: proposal.call_data.num_coins_to_forward, rC: proposal.call_data.asset_id_of_coins_to_forward, rD: proposal.call_data.amount_of_gas_to_forward) {
+            call rA rB rC rD;
+        }
+
+        // Users can now convert their votes back into tokens
```

Comment:
SHould they be able to though?

Reply:
Why should they not be able to?  After a proposal expires and can no longer be voted on they should be able to "unlock" their votes if they want to withdraw their tokens.

Reply:
It could be a disincentive to prevent the spam of crappy proposals. Winning proposals are unlocked, crappy ones get burned. Design decision... I don't think it matters in the first iteration. More thought can be given to expanding the logic or creating modules which allow you to switch in other logic modules later on

Reply:
Yes there is still work to do on how a proposal gets added.  Iirc the Maker DAO system for adding proposals is basically: only specific accounts can add proposals, and that's how they prevent spam.